bionemo-recipes/models/esm2/tests/test_cp.py

```diff
@@ -0,0 +1,396 @@
+        cp_rank: The context parallelism rank.
+        cp_world_size: The size of the context parallelism group.
+    Returns:
+        A dictionary containing the shard of the batch.
+    """
+    input_ids_sharded, labels_sharded = split_batch_by_cp_rank(
+        cu_seqlens_padded=batch["cu_seq_lens_q_padded"],
+        input_ids_padded=batch["input_ids"],
+        labels_padded=batch["labels"],
+        qvk_format="thd",
+        cp_rank=cp_rank,
+        cp_world_size=cp_world_size,
+    )
+    batch_shard = dict(batch)
+    batch_shard["input_ids"] = input_ids_sharded
+    batch_shard["labels"] = labels_sharded
+    # Now determine the max length of the sequence.
+    seqlens_q = batch_shard["cu_seq_lens_q_padded"][1:] - batch_shard["cu_seq_lens_q_padded"][:-1]
+    batch_shard["max_length_q"] = int((seqlens_q.max().item() + 63) // 64 * 64)  # From TE code.
+    batch_shard["max_length_k"] = batch_shard["max_length_q"]
+    return batch_shard
+
+
+@dataclass(frozen=True)
+class DistributedConfig:
+    """Class to track distributed ranks and handle basic distributed training setup.
+
+    If torch distributed environment variables are not set, we set them to default values for single-process training.
+
+    Attributes:
+        rank: The rank of the process.
+        local_rank: The local rank of the process.
+        world_size: The total number of processes.
+    """
+
+    rank: int = field(default_factory=lambda: int(os.environ.setdefault("RANK", "0")))
+    local_rank: int = field(default_factory=lambda: int(os.environ.setdefault("LOCAL_RANK", "0")))
+    world_size: int = field(default_factory=lambda: int(os.environ.setdefault("WORLD_SIZE", "1")))
+    _master_addr: str = field(default_factory=lambda: os.environ.setdefault("MASTER_ADDR", "localhost"))
+    _master_port: str = field(default_factory=lambda: os.environ.setdefault("MASTER_PORT", "12355"))
+
+    def is_main_process(self) -> bool:
+        """This is the global rank 0 process, to be used for wandb logging, etc."""
+        return self.rank == 0
+
+
+@requires_multi_gpu
+def test_context_parallel_equivalence_2process():
+    """
+    Test the context parallel equivalence between 2 processes. In one instance, we run the model in non-distributed mode and in the other
+    we run the model in distributed mode with context parallelism. We then compare the losses and logits from the two runs.
+
+    We compare the (1) Losses, (2) Logits, and (3) Gradients from the two runs.
+    """
+    cmd = [
+        "torchrun",
+        "--nproc_per_node=2",
+        os.path.relpath(__file__),
+    ]
+    result = subprocess.run(
+        cmd,
+        check=False,
+        text=True,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        timeout=240,
+    )
+    if result.returncode != 0:
+        print(f"STDOUT:\n{result.stdout}")
+        print(f"STDERR:\n{result.stderr}")
+        pytest.fail(f"Command failed with exit code {result.returncode}")
```

Comment:
this file won't run at all on our 1-gpu CI nodes -- is it worth it at all to run a 1-GPU version just to make sure these functions all fire correctly?